### DISCUSSION TITLE
control plane for nikss

### DISCUSSION THREAD

**Discussion by Arishbiswas** [NIKSS-vSwitch/nikss#114] [Q&A]
currently nikss doesnot have p4runtime integration, in that case is there any way possible that nikss switch can be integrated with any external controller. If so, how to do it?

**Comment by tatry** [comment]
> currently nikss doesnot have p4runtime integration

That is not entirely true. Some time we started to add support for P4Runtime, but for some reason work has been stopped. Our plan was integration with Stratum. [Our fork of Stratum](https://github.com/NIKSS-vSwitch/stratum) has some initial support for NIKSS. There is [another fork](https://github.com/Kinnyeo/stratum) with wider support for NIKSS (like adding/removing table entries) but has not been merged yet into our Stratum fork.

> is there any way possible that nikss switch can be integrated with any external controller. If so, how to do it?

Yes, it is possible. There are at least two possibilities to do it.

The best way I know is to use C API provided by libnikss (part of `nikss` repository, make sure to compile with option `BUILD_SHARED` set to `on` - then you will have separate .so lib that you can link against your binary). As a side note you can see `nikss-ctl` as some sort of controller here.

If you are using some scripting language (like python), it might simpler to build command line for `nikss-ctl` instead of calling C API. We have done [something like that](https://github.com/p4lang/p4c/blob/main/backends/ebpf/tests/ptf/common.py) for testing p4c compiler (aka PTF tests).

**Comment by Arishbiswas** [comment]
Thankyou for your reply. but, my problem is, i am doing a project where i need the controller messages to be recorded , like suppose in a pcap file or in a log packet which should be used be used for further nanlysis. is that possible in case of nikss-ctl. if so how?

**Comment by tatry** [comment]
I would say it depends on your project. But in any rate, I think there is no ready to use solution for now.

One way would be create simple wrapper for `nikss-ctl` that records command line in log file and then call original `nikss-ctl` (even bash script could be fine if I'm correct). Or some additional C code in `nikss-ctl` that records command line arguments.

Another way would be integration with controller and usage of it for your job. Unfortunately such integration is not ready yet, especially with Stratum which could be fine for you.

Maybe there are other possibilities but now I'm not aware of them.
